pkg/agent/openflow/network_policy.go

```diff
@@ -2220,3 +2227,69 @@ func (f *featureNetworkPolicy) l7NPTrafficControlFlows() []binding.Flow {
 			Done(),
 	}
 }
+
+func (f *featureNetworkPolicy) loggingNPPacketFlowWithOperations(cookieID uint64, operations uint8) binding.Flow {
+	loggingOperations := binding.NewRegMark(PacketInOperationField, uint32(operations))
+	fb := L2ForwardingOutTable.ofTable.BuildFlow(priorityNormal).
+		MatchRegMark(SendToControllerRegMark, loggingOperations)
+	if f.ovsMetersAreSupported {
+		fb = fb.Action().Meter(PacketInMeterIDNP)
+	}
+	return fb.Action().SendToController([]byte{uint8(PacketInCategoryNP), operations}, false).
+		Cookie(cookieID).
+		Done()
+}
+
+func (f *featureNetworkPolicy) initLoggingFlows() []binding.Flow {
+	if !f.enableAntreaPolicy {
+		return nil
+	}
+	maxOperationValue := PacketInNPLoggingOperation + PacketInNPStoreDenyOperation + PacketInNPRejectOperation
+	flows := make([]binding.Flow, 0, maxOperationValue)
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	for operation := 1; operation <= maxOperationValue; operation++ {
+		flows = append(flows, f.loggingNPPacketFlowWithOperations(cookieID, uint8(operation)))
+	}
+	return flows
+}
+
+func (f *featureNetworkPolicy) loggingGroups() []binding.OFEntry {
+	var groups []binding.OFEntry
+	if !f.enableAntreaPolicy {
+		return groups
+	}
+	candidateTables := []*Table{EgressRuleTable, EgressMetricTable, IngressRuleTable, IngressMetricTable}
+	if f.enableMulticast {
+		candidateTables = append(candidateTables, MulticastEgressMetricTable, MulticastIngressMetricTable)
+	}
+	for _, nextTable := range candidateTables {
+		groupKey := fmt.Sprintf("%d", nextTable.GetID())
+		obj, ok := f.loggingGroupCache.Load(groupKey)
+		if ok {
+			groups = append(groups, obj.(binding.Group))
+			continue
+		}
+		// Create OpenFlow group to both log Antrea-native policy events and resubmit the packet to nextTable.
+		groupID := f.groupAllocator.Allocate()
+		// There are two buckets in this type All group which have generated two copies of the packet: 1) resubmit
+		// the first one back to the next table of which the original packet consumes the group, then it is able to
+		// continue with the previous forwarding logic; 2) set packetIn marks in the second one and resubmit it to
+		// L2ForwardingOutTable.
+		group := f.bridge.NewGroupTypeAll(groupID).Bucket().
+			ResubmitToTable(nextTable.GetID()).
+			Done().
+			Bucket().
+			LoadRegMark(SendToControllerRegMark).
+			ResubmitToTable(L2ForwardingOutTable.GetID()).
```

Comment:
I feel reusing L2ForwardingOutTable for sending to controller may introduce some confusion, do you think we can just use a separate table for this?

Reply:
In my opinion, L2ForwardingOutTable is working as the last table where packets will leave the OVS pipeline, the ways for a packet to leave the OVS pipeline include 1) output to a port; 2) send to OpenFlow controller (antrea-agent). So I didn't introduce new table in the changes.